scheduler/schedule_executor.go

```diff
@@ -0,0 +1,111 @@
+package scheduler
+
+import (
+	"context"
+	"time"
+
+	runtimeInterfaces "github.com/flyteorg/flyteadmin/pkg/runtime/interfaces"
+	"github.com/flyteorg/flyteadmin/scheduler/core"
+	"github.com/flyteorg/flyteadmin/scheduler/executor"
+	"github.com/flyteorg/flyteadmin/scheduler/repositories"
+	"github.com/flyteorg/flyteadmin/scheduler/snapshoter"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/service"
+	"github.com/flyteorg/flytestdlib/futures"
+	"github.com/flyteorg/flytestdlib/logger"
+	"github.com/flyteorg/flytestdlib/promutils"
+
+	"golang.org/x/time/rate"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+const snapshotWriterDuration = 30 * time.Second
+const scheduleUpdaterDuration = 30 * time.Second
+
+const snapShotVersion = 1
+
+// ScheduledExecutor used for executing the schedules saved by the native flyte scheduler in the database.
+type ScheduledExecutor struct {
+	scheduler              core.Scheduler
+	snapshoter             snapshoter.Persistence
+	db                     repositories.SchedulerRepoInterface
+	scope                  promutils.Scope
+	adminServiceClient     service.AdminServiceClient
+	workflowExecutorConfig *runtimeInterfaces.FlyteWorkflowExecutorConfig
+}
+
+func (w *ScheduledExecutor) Run(ctx context.Context) error {
+	logger.Infof(ctx, "Flyte native scheduler started successfully")
+
+	defer logger.Infof(ctx, "Flyte native scheduler shutdown")
+
+	// Read snapshot from the DB.
+	snapShotReader := &snapshoter.VersionedSnapshot{Version: snapShotVersion}
+	snapshot, err := w.snapshoter.Read(ctx, snapShotReader)
+
+	if err != nil {
+		logger.Errorf(ctx, "unable to read the snapshot from the db due to %v. Aborting", err)
+		return err
+	}
+
+	// Read all the schedules from the DB
+	schedules, err := w.db.SchedulableEntityRepo().GetAll(ctx)
+	if err != nil {
+		logger.Errorf(ctx, "unable to read the schedules from the db due to %v. Aborting", err)
+		return err
+	}
+	adminRateLimit := w.workflowExecutorConfig.GetAdminRateLimit()
+
+	// Set the rate limit on the admin
+	rateLimiter := rate.NewLimiter(adminRateLimit.GetTps(), adminRateLimit.GetBurst())
+
+	// Set the executor to send executions to admin
+	executor := executor.New(w.scope, w.adminServiceClient)
+
+	// Create the scheduler using GoCronScheduler implementation
+	gcronScheduler := core.NewGoCronScheduler(w.scope, snapshot, rateLimiter, executor)
+	w.scheduler = gcronScheduler
+
+	// Bootstrap the schedules from the snapshot
+	bootStrapCtx, bootStrapCancel := context.WithCancel(ctx)
+	defer bootStrapCancel()
+	gcronScheduler.BootStrapSchedulesFromSnapShot(bootStrapCtx, schedules, snapshot)
```

Comment:
Instead of bootstrap can we make this part of the constructor?

Reply:
as in part of the NewGoCron... method

Reply:
Done